File: battleship.rb
# battleship.rb
# Версия на Ruby с метапрограммированием, блоками, цветным выводом

# ANSI colors
COLORS = {
  reset: "\e[0m",
  red: "\e[91m",
  green: "\e[92m",
  blue: "\e[94m",
  yellow: "\e[93m",
  cyan: "\e[96m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

class Ship
  attr_reader :cells

  def initialize(cells)
    @cells = cells
    @hits = Array.new(cells.size, false)
  end

  def sunk?
    @hits.all?
  end

  def hit?(row, col)
    @cells.each_with_index do |(r, c), i|
      if r == row && c == col
        @hits[i] = true
        return true
      end
    end
    false
  end
end

class Board
  SIZE = 10
  SHIP_SIZES = [4, 3, 3, 2, 2, 2, 1, 1, 1, 1]

  attr_reader :shots

  def initialize
    @grid = Array.new(SIZE) { Array.new(SIZE, '~') }
    @ships = []
    @shots = Set.new
  end

  def place_ships
    SHIP_SIZES.each do |size|
      placed = false
      attempts = 0
      until placed || attempts >= 1000
        attempts += 1
        row = rand(SIZE)
        col = rand(SIZE)
        horizontal = [true, false].sample
        cells = get_cells(row, col, size, horizontal)
        if cells && can_place?(cells)
          place_ship(cells)
          placed = true
        end
      end
      raise "Не удалось разместить корабли" unless placed
    end
  end

  private

  def get_cells(row, col, size, horizontal)
    cells = []
    size.times do |i|
      r = horizontal ? row : row + i
      c = horizontal ? col + i : col
      return nil if r >= SIZE || c >= SIZE
      cells << [r, c]
    end
    cells
  end

  def can_place?(cells)
    cells.all? do |r, c|
      return false unless @grid[r][c] == '~'
      (-1..1).all? do |dr|
        (-1..1).all? do |dc|
          nr = r + dr
          nc = c + dc
          next true if nr < 0 || nr >= SIZE || nc < 0 || nc >= SIZE
          @grid[nr][nc] == '~'
        end
      end
    end
  end

  def place_ship(cells)
    ship = Ship.new(cells)
    @ships << ship
    cells.each { |r, c| @grid[r][c] = '#' }
  end

  public

  def receive_shot(row, col)
    return 'invalid' if row < 0 || row >= SIZE || col < 0 || col >= SIZE
    return 'already_shot' if @shots.include?([row, col])
    @shots.add([row, col])

    if @grid[row][col] == '#'
      @ships.each do |ship|
        if ship.hit?(row, col)
          @grid[row][col] = 'X'
          if ship.sunk?
            ship.cells.each { |r, c| @grid[r][c] = 'X' }
            return 'sunk'
          end
          return 'hit'
        end
      end
    else
      @grid[row][col] = 'O'
      return 'miss'
    end
    'miss'
  end

  def all_ships_sunk?
    @ships.all?(&:sunk?)
  end

  def display(hide_ships: false)
    print '  '
    (1..SIZE).each { |i| print "#{i} " }
    puts
    SIZE.times do |r|
      row_label = ('A'.ord + r).chr
      print "#{row_label} "
      SIZE.times do |c|
        cell = @grid[r][c]
        cell = '~' if hide_ships && cell == '#'
        colored = case cell
                  when 'X' then colorize('X', :red)
                  when 'O' then colorize('O', :blue)
                  when '#' then colorize('#', :green)
                  else cell
                  end
        print "#{colored} "
      end
      puts
    end
  end
end

class Game
  def initialize
    @player_board = Board.new
    @computer_board = Board.new
    @player_board.place_ships
    @computer_board.place_ships
    @player_turn = true
    @game_over = false
  end

  def parse_coordinate(input)
    match = input.match(/^([A-Ja-j])([1-9]|10)$/)
    return nil unless match
    row = match[1].upcase.ord - 'A'.ord
    col = match[2].to_i - 1
    [row, col]
  end

  def player_move
    puts "\nВаше поле:"
    @player_board.display(hide_ships: false)
    puts "\nПоле противника:"
    @computer_board.display(hide_ships: true)

    loop do
      print "Введите координату (например, A1): "
      input = gets.chomp.strip
      parsed = parse_coordinate(input)
      unless parsed
        puts "Неверный формат. Используйте букву A-J и цифру 1-10."
        next
      end
      row, col = parsed
      result = @computer_board.receive_shot(row, col)
      case result
      when 'invalid' then puts "Координата вне поля."
      when 'already_shot' then puts "Сюда уже стреляли."
      when 'hit' then puts colorize("Попадание!", :green)
      when 'sunk' then puts colorize("Корабль потоплен!", :yellow)
      when 'miss' then puts colorize("Промах.", :blue)
      end
      break unless ['invalid', 'already_shot'].include?(result)
    end

    if @computer_board.all_ships_sunk?
      puts colorize("Поздравляем! Вы потопили все корабли противника!", :green)
      @game_over = true
    end
  end

  def computer_move
    puts "\nХод компьютера..."
    loop do
      row = rand(Board::SIZE)
      col = rand(Board::SIZE)
      unless @player_board.shots.include?([row, col])
        result = @player_board.receive_shot(row, col)
        coord_str = ('A'.ord + row).chr + (col + 1).to_s
        if ['hit', 'sunk'].include?(result)
          puts "Компьютер попал в #{coord_str}!"
        else
          puts "Компьютер промахнулся по #{coord_str}."
        end
        break
      end
    end

    if @player_board.all_ships_sunk?
      puts colorize("Компьютер уничтожил все ваши корабли. Вы проиграли.", :red)
      @game_over = true
    end
  end

  def run
    puts colorize("Добро пожаловать в Морской бой!", :cyan)
    until @game_over
      if @player_turn
        player_move
      else
        computer_move
      end
      @player_turn = !@player_turn
    end
  end
end

Game.new.run

File: battleship.py
# battleship.py
# Версия на Python с dataclasses, type hints, цветным выводом (colorama не используется, используем ANSI-коды)

import random
import sys
import re
from dataclasses import dataclass
from typing import List, Tuple, Optional

# ANSI-цвета для терминала
COLORS = {
    'reset': '\033[0m',
    'red': '\033[91m',
    'green': '\033[92m',
    'blue': '\033[94m',
    'yellow': '\033[93m',
    'cyan': '\033[96m',
}


def colorize(text: str, color: str) -> str:
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"


@dataclass
class Ship:
    """Класс корабля."""
    cells: List[Tuple[int, int]]
    hits: List[bool]  # список попаданий по каждой клетке

    def is_sunk(self) -> bool:
        return all(self.hits)

    def hit(self, row: int, col: int) -> bool:
        for i, (r, c) in enumerate(self.cells):
            if r == row and c == col:
                self.hits[i] = True
                return True
        return False


class Board:
    """Игровое поле."""
    SIZE = 10
    SHIP_SIZES = [4, 3, 3, 2, 2, 2, 1, 1, 1, 1]

    def __init__(self):
        self.grid = [['~' for _ in range(self.SIZE)] for _ in range(self.SIZE)]
        self.ships: List[Ship] = []
        self.shots: List[Tuple[int, int]] = []  # все выстрелы по этому полю

    def place_ships(self) -> None:
        """Случайно расставляет корабли."""
        for size in self.SHIP_SIZES:
            placed = False
            attempts = 0
            while not placed and attempts < 1000:
                attempts += 1
                row = random.randint(0, self.SIZE - 1)
                col = random.randint(0, self.SIZE - 1)
                horizontal = random.choice([True, False])
                cells = self._get_cells(row, col, size, horizontal)
                if cells and self._can_place(cells):
                    self._place_ship(cells)
                    placed = True
            if not placed:
                raise RuntimeError("Не удалось разместить корабли")

    def _get_cells(self, row: int, col: int, size: int, horizontal: bool) -> Optional[List[Tuple[int, int]]]:
        cells = []
        for i in range(size):
            r = row if horizontal else row + i
            c = col + i if horizontal else col
            if r >= self.SIZE or c >= self.SIZE:
                return None
            cells.append((r, c))
        return cells

    def _can_place(self, cells: List[Tuple[int, int]]) -> bool:
        for r, c in cells:
            if self.grid[r][c] != '~':
                return False
            # Проверка соседей (чтобы корабли не касались)
            for dr in (-1, 0, 1):
                for dc in (-1, 0, 1):
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < self.SIZE and 0 <= nc < self.SIZE:
                        if self.grid[nr][nc] != '~':
                            return False
        return True

    def _place_ship(self, cells: List[Tuple[int, int]]) -> None:
        ship = Ship(cells, [False] * len(cells))
        self.ships.append(ship)
        for r, c in cells:
            self.grid[r][c] = '#'

    def receive_shot(self, row: int, col: int) -> str:
        """Обрабатывает выстрел по полю. Возвращает 'hit', 'miss', 'sunk', 'invalid'."""
        if not (0 <= row < self.SIZE and 0 <= col < self.SIZE):
            return 'invalid'
        if (row, col) in self.shots:
            return 'already_shot'
        self.shots.append((row, col))
        if self.grid[row][col] == '#':
            # Попадание
            for ship in self.ships:
                if ship.hit(row, col):
                    self.grid[row][col] = 'X'
                    if ship.is_sunk():
                        # Отметить все клетки корабля как X
                        for r, c in ship.cells:
                            self.grid[r][c] = 'X'
                        return 'sunk'
                    return 'hit'
        else:
            self.grid[row][col] = 'O'
            return 'miss'
        return 'miss'  # fallback

    def all_ships_sunk(self) -> bool:
        return all(ship.is_sunk() for ship in self.ships)

    def display(self, hide_ships: bool = False) -> None:
        """Вывод поля."""
        print('  ' + ' '.join(str(i+1) for i in range(self.SIZE)))
        for r in range(self.SIZE):
            row_label = chr(ord('A') + r)
            row_cells = []
            for c in range(self.SIZE):
                cell = self.grid[r][c]
                if hide_ships and cell == '#':
                    cell = '~'
                if cell == 'X':
                    row_cells.append(colorize('X', 'red'))
                elif cell == 'O':
                    row_cells.append(colorize('O', 'blue'))
                elif cell == '#':
                    row_cells.append(colorize('#', 'green'))
                else:
                    row_cells.append(cell)
            print(f"{row_label} " + ' '.join(row_cells))


class BattleshipGame:
    def __init__(self):
        self.player_board = Board()
        self.computer_board = Board()
        self.player_board.place_ships()
        self.computer_board.place_ships()
        self.player_turn = True
        self.game_over = False

    def parse_coordinate(self, coord: str) -> Optional[Tuple[int, int]]:
        """Преобразует 'A1' в (row, col)."""
        match = re.match(r'^([A-Ja-j])([1-9]|10)$', coord)
        if not match:
            return None
        row = ord(match.group(1).upper()) - ord('A')
        col = int(match.group(2)) - 1
        return row, col

    def player_move(self) -> None:
        print("\nВаше поле:")
        self.player_board.display(hide_ships=False)
        print("\nПоле противника:")
        self.computer_board.display(hide_ships=True)

        while True:
            coord = input("Введите координату (например, A1): ").strip()
            parsed = self.parse_coordinate(coord)
            if parsed is None:
                print("Неверный формат. Используйте букву A-J и цифру 1-10.")
                continue
            row, col = parsed
            result = self.computer_board.receive_shot(row, col)
            if result == 'invalid':
                print("Координата вне поля.")
            elif result == 'already_shot':
                print("Сюда уже стреляли.")
            else:
                if result == 'hit':
                    print(colorize("Попадание!", 'green'))
                elif result == 'sunk':
                    print(colorize("Корабль потоплен!", 'yellow'))
                elif result == 'miss':
                    print(colorize("Промах.", 'blue'))
                break

        # Проверка победы
        if self.computer_board.all_ships_sunk():
            print(colorize("Поздравляем! Вы потопили все корабли противника!", 'green'))
            self.game_over = True

    def computer_move(self) -> None:
        print("\nХод компьютера...")
        # Случайный выстрел
        while True:
            row = random.randint(0, Board.SIZE - 1)
            col = random.randint(0, Board.SIZE - 1)
            if (row, col) not in self.player_board.shots:
                break
        result = self.player_board.receive_shot(row, col)
        if result == 'hit' or result == 'sunk':
            print(f"Компьютер попал в {chr(ord('A')+row)}{col+1}!")
        else:
            print(f"Компьютер промахнулся по {chr(ord('A')+row)}{col+1}.")

        if self.player_board.all_ships_sunk():
            print(colorize("Компьютер уничтожил все ваши корабли. Вы проиграли.", 'red'))
            self.game_over = True

    def run(self) -> None:
        print(colorize("Добро пожаловать в Морской бой!", 'cyan'))
        while not self.game_over:
            if self.player_turn:
                self.player_move()
            else:
                self.computer_move()
            self.player_turn = not self.player_turn


if __name__ == '__main__':
    game = BattleshipGame()
    game.run()
